File: Tests/ConnectorControlCoreTests/PasteRecoveryTests.swift
import XCTest
@testable import ConnectorControlCore

final class PasteRecoveryTests: XCTestCase {
    private func command(_ v: JSONValue?) -> String? {
        guard case .object(let o)? = v, case .string(let s)? = o["command"] else { return nil }
        return s
    }

    func testBareStanzaWithTrailingBrace() {
        // Exactly the shape copied out of an mcpServers block: a bare
        // "name": {…} property plus the block's leftover closing brace.
        let text = """
          "okta-mcp-server": {
            "command": "/opt/homebrew/bin/uv",
            "args": ["run", "--directory", "/x", "okta-mcp-server"],
            "env": { "OKTA_ORG_URL": "https://example.okta.com" }
          }
        }
        """
        let r = PasteRecovery.recover(text)
        XCTAssertEqual(r?.name, "okta-mcp-server")
        XCTAssertEqual(command(r?.config), "/opt/homebrew/bin/uv")
    }

    func testBareStanzaWithoutTrailingBrace() {
        let text = "\"foo\": {\"command\": \"npx\"}"
        let r = PasteRecovery.recover(text)
        XCTAssertEqual(r?.name, "foo")
        XCTAssertEqual(command(r?.config), "npx")
    }

    func testMcpServersWrapper() {
        let text = "{\"mcpServers\": {\"bar\": {\"command\": \"uvx\"}}}"
        let r = PasteRecovery.recover(text)
        XCTAssertEqual(r?.name, "bar")
        XCTAssertEqual(command(r?.config), "uvx")
    }

    func testSingleEntryNameWrapper() {
        let text = "{\"baz\": {\"command\": \"node\"}}"
        let r = PasteRecovery.recover(text)
        XCTAssertEqual(r?.name, "baz")
        XCTAssertEqual(command(r?.config), "node")
    }

    func testPlainConfigObjectIsNotRenamed() {
        let text = "{\"command\": \"npx\", \"args\": [\"-y\", \"pkg\"]}"
        let r = PasteRecovery.recover(text)
        XCTAssertNil(r?.name, "a bare config must not be treated as name-wrapped")
        XCTAssertEqual(command(r?.config), "npx")
    }

    func testSingleKeyConfigNotUnwrapped() {
        // {"command": "x"} is a one-key CONFIG, not a {name: config} wrapper.
        let r = PasteRecovery.recover("{\"command\": \"x\"}")
        XCTAssertNil(r?.name)
        XCTAssertEqual(command(r?.config), "x")
    }

    func testBracesInsideStringValuesAreIgnored() {
        let text = "\"weird\": {\"command\": \"echo }}}\"}"
        let r = PasteRecovery.recover(text)
        XCTAssertEqual(r?.name, "weird")
        XCTAssertEqual(command(r?.config), "echo }}}")
    }

    func testMultiEntryMcpServersNotMisnamed() {
        let text = "{\"mcpServers\": {\"a\": {\"command\":\"x\"}, \"b\": {\"command\":\"y\"}}}"
        let r = PasteRecovery.recover(text)
        XCTAssertNotEqual(r?.name, "mcpServers",
                          "a multi-entry wrapper must not be unwrapped to name=mcpServers")
    }

    func testGarbageReturnsNil() {
        XCTAssertNil(PasteRecovery.recover("{{{ not json"))
        XCTAssertNil(PasteRecovery.recover("   "))
    }
}

File: windows/tests/ConnectorControl.Core.Tests/PasteRecoveryTests.cs
namespace ConnectorControl.Core.Tests;

public class PasteRecoveryTests
{
    private static string? Command(JsonValue? v) => v?["command"] is { Kind: JsonKind.String } s ? s.StringValue : null;

    private static string[]? Args(JsonValue? v) =>
        v?["args"] is { Kind: JsonKind.Array } a ? a.ArrayItems.Where(i => i.Kind == JsonKind.String).Select(i => i.StringValue).ToArray() : null;

    [Fact]
    public void BareStanzaWithTrailingBrace()
    {
        const string text = "  \"okta-mcp-server\": {\n    \"command\": \"/opt/homebrew/bin/uv\",\n    \"args\": [\"run\", \"--directory\", \"/x\", \"okta-mcp-server\"],\n    \"env\": { \"OKTA_ORG_URL\": \"https://example.okta.com\" }\n  }\n}\n";
        var r = PasteRecovery.Recover(text);
        Assert.Equal("okta-mcp-server", r?.Name);
        Assert.Equal("/opt/homebrew/bin/uv", Command(r?.Config));
    }

    [Fact]
    public void BareStanzaWithoutTrailingBrace()
    {
        var r = PasteRecovery.Recover("\"foo\": {\"command\": \"npx\"}");
        Assert.Equal("foo", r?.Name);
        Assert.Equal("npx", Command(r?.Config));
    }

    [Fact]
    public void McpServersWrapper()
    {
        var r = PasteRecovery.Recover("{\"mcpServers\": {\"bar\": {\"command\": \"uvx\"}}}");
        Assert.Equal("bar", r?.Name);
        Assert.Equal("uvx", Command(r?.Config));
    }

    [Fact]
    public void SingleEntryNameWrapper()
    {
        var r = PasteRecovery.Recover("{\"baz\": {\"command\": \"node\"}}");
        Assert.Equal("baz", r?.Name);
        Assert.Equal("node", Command(r?.Config));
    }

    [Fact]
    public void PlainConfigObjectIsNotRenamed()
    {
        var r = PasteRecovery.Recover("{\"command\": \"npx\", \"args\": [\"-y\", \"pkg\"]}");
        Assert.NotNull(r);
        Assert.Null(r.Name);
        Assert.Equal("npx", Command(r.Config));
    }

    [Fact]
    public void SingleKeyConfigNotUnwrapped()
    {
        var r = PasteRecovery.Recover("{\"command\": \"x\"}");
        Assert.NotNull(r);
        Assert.Null(r.Name);
        Assert.Equal("x", Command(r.Config));
    }

    [Fact]
    public void BracesInsideStringValuesAreIgnored()
    {
        var r = PasteRecovery.Recover("\"weird\": {\"command\": \"echo }}}\"}");
        Assert.Equal("weird", r?.Name);
        Assert.Equal("echo }}}", Command(r?.Config));
    }

    [Fact]
    public void MultiEntryMcpServersNotMisnamed()
    {
        var r = PasteRecovery.Recover("{\"mcpServers\": {\"a\": {\"command\":\"x\"}, \"b\": {\"command\":\"y\"}}}");
        Assert.NotEqual("mcpServers", r?.Name);
    }

    [Fact]
    public void GarbageReturnsNull()
    {
        Assert.Null(PasteRecovery.Recover("{{{ not json"));
        Assert.Null(PasteRecovery.Recover("   "));
    }

    [Fact]
    public void CurlyClosingQuoteIsNormalized()
    {
        var r = PasteRecovery.Recover("\"splunk\": {\"command\": \"npx\", \"args\": [\"-y\", \"mcp-remote@latest”]}");
        Assert.Equal("splunk", r?.Name);
        Assert.Equal("npx", Command(r?.Config));
        Assert.Equal(["-y", "mcp-remote@latest"], Args(r?.Config)!);
    }

    [Fact]
    public void ValidStraightQuoteConfigIsUnaffectedByNormalization()
    {
        var r = PasteRecovery.Recover("{\"command\": \"npx\", \"args\": [\"-y\", \"pkg\"]}");
        Assert.NotNull(r);
        Assert.Null(r.Name);
        Assert.Equal("npx", Command(r.Config));
        Assert.Equal(["-y", "pkg"], Args(r.Config)!);
    }
}
